rust/src/certificate.rs

```diff
@@ -1,49 +1,54 @@
-use crate::PetError;
+use derive_more::{AsMut, AsRef};
 
-#[derive(Debug, PartialEq)]
+use crate::{crypto::ByteObject, PetError};
+
+#[derive(AsRef, AsMut, Clone, Debug, PartialEq)]
 /// A dummy certificate.
 pub struct Certificate(Vec<u8>);
 
-#[allow(clippy::len_without_is_empty)]
-impl Certificate {
-    #[allow(clippy::new_without_default)]
-    /// Create a certificate
-    pub fn new() -> Self {
-        Self(vec![0_u8; 32])
+impl ByteObject for Certificate {
+    /// Create a certificate a slice of bytes. Fails if the length of the input is invalid.
+    fn from_slice(bytes: &[u8]) -> Option<Self> {
+        Some(Self(bytes.to_vec()))
     }
 
-    /// Get the length of the certificate.
-    pub fn len(&self) -> usize {
-        self.as_ref().len()
+    /// Create a certificate initialized to zero.
+    fn zeroed() -> Self {
+        Self(vec![0_u8; Self::BYTES])
     }
 
-    /// Validate a certificate
-    pub fn validate(&self) -> Result<(), PetError> {
-        if self.as_ref() == [0_u8; 32] {
-            Ok(())
-        } else {
-            Err(PetError::InvalidMessage)
-        }
+    /// Get the certificate as a slice.
+    fn as_slice(&self) -> &[u8] {
+        self.0.as_slice()
     }
 }
 
-impl AsRef<[u8]> for Certificate {
-    /// Get a reference to the certificate.
-    fn as_ref(&self) -> &[u8] {
-        self.0.as_slice()
+#[allow(clippy::len_without_is_empty)]
+impl Certificate {
+    /// Get the number of bytes of a certificate.
+    pub const BYTES: usize = 32;
```

Comment:
fwiw I named that `LENGTH` in https://github.com/xainag/xain-fl/pull/398. I think it's a bit clearer. But does it make sense to have this for certificates? Their size is not fixed.

Reply:
i've seen the changes for `LENGTH`, makes sense to adopt this. regarding this constant for certificates, it's just a dummy for now and will probably change once we implement/use real certificates.